Literal check on config section and key by mypy

sakurai-youhei opened the pull request:
This PR will introduce a type check by mypy, which helps find illegal literals on the config section and key.

Relates #1646

gbanasiak commented on pyproject.toml:
```diff
@@ -139,3 +139,11 @@ junit_family = "xunit2"
 junit_logging = "all"
 asyncio_mode = "strict"
 xfail_strict = true
+
+[tool.mypy]
+python_version = 3.8
+disable_error_code = [
+    "import-not-found",
+    "import-untyped",
+    "var-annotated",
+]
```
```suggestion

# With rare exceptions, Rally does not use type hints. The intention of the
# following largely reduced mypy configuration scope is verification of argument
# types in config.Config methods while introducing configuration properties
# (props). The error we are after here is "arg-type".
[tool.mypy]
python_version = 3.8
check_untyped_defs = true
disable_error_code = [
    "assignment",
    "attr-defined",
    "call-overload",
    "dict-item",
    "import-not-found",
    "import-untyped",
    "index",
    "misc",
    "operator",
    "str-bytes-safe",
    "union-attr",
    "var-annotated",
]
```

gbanasiak reviewed (changes requested):
I like the idea of using mypy to make sure we covered all config sections/keys. At the same time we don't want to turn this into type hints introduction project for Rally. I don't think the effort is justified. I think we should aim at reducing the scope of this PR to absolute minimum. For instance modifications in `esrally/client/synchronous.py` and `esrally/client/asynchronous.py` are not relevant, so I think should be reverted.

I've played with this setup a bit and think that if we narrow down mypy to react to `arg-type` error _only_ we should achieve the verification we want here, i.e. verification of arguments of `config.Config` class methods (see suggestion). If `arg-type` error is triggered for anything else than `config.Config` methods I would add `# type: ignore[arg-type]` annotation.

I've noticed that the list of literals is not complete due to lack of `check_untyped_defs = true` option for mypy.

sakurai-youhei commented:
@gbanasiak Thank you for your comments. Please feel free to bring up anything.
- I agree to minimize the scope.
- I added `# type: ignore[arg-type]` with TODO comment.
- I didn't notice `check_untyped_defs = true` is needed. 👍 
- I also added tests to check the literals to avoid human errors.

gbanasiak reviewed (commented):
Thank you for the next iteration. I really like the idea with unit tests :)

I did the following sanity check:
```
% ipython
[..]
In [1]: from esrally.config import Section, Key

In [2]: len(str(Section).split(","))
Out[2]: 21

In [3]: len(str(Key).split(","))
Out[3]: 105
```

Then in shell:
```
rg -oN --heading "g\.add\(config\.Scope\..*?, \".*?\", " | grep -Eo "\".*?\"" | sort -u > /tmp/section-tmp.txt
rg -oN --heading "g\.opts\(\".*?\"," | grep -Eo "\".*?\"" | sort -u >> /tmp/section-tmp.txt
sort -u /tmp/section-tmp.txt > /tmp/section.txt
wc -l /tmp/section.txt
rm /tmp/section-tmp.txt

rg -No --heading "g\.add\(config\.Scope\..*?, (\".*?\", ){2}" | awk '{print $3;}' | grep -Eo "\".*?\"" | sort -u > /tmp/key-tmp.txt
rg -oN --heading "g\.opts\(\".*?\)" | awk '{print $2;}' | grep -Eo "\".*?\"" | sort -u >> /tmp/key-tmp.txt
sort -u /tmp/key-tmp.txt > /tmp/key.txt
wc -l /tmp/key.txt
rm /tmp/key-tmp.txt
```

which produced:
```
      21 /tmp/section.txt
     115 /tmp/key.txt
```

Based on this I think we are right on target with the sections, but some of the keys are still missing. Just as an example (you can reproduce the above for complete list), the first missing key was `distribution.dir` which is present in `esrally/mechanic/supplier.py`.

Can you revisit the missing keys?

sakurai-youhei commented on esrally/driver/driver.py:
```diff
@@ -404,7 +404,7 @@ def __init__(self):
         self.logger = logging.getLogger(__name__)
         self.track_name = None
-        self.cfg = None
+        self.cfg: Union[config.Config, None] = None
```
I feel it's more natural to type the `self.cfg` to `None` and leave room for `config.Config`. And the `self.cfg` is not something like an optional argument. But based on this theory, even the current use of Union looks unnatural now.

> https://docs.python.org/3/library/typing.html#typing.Union
> `Union[X, Y]` is equivalent to `X | Y` and means either X or Y.

My best would be `Union[None, config.Config]`. I seem to prefer explicitly over simplicity. But does my preference make the code hard to read?

Please let me know which is more welcome in the Rally project, and I will follow the ideea. @gbanasiak

gbanasiak replied:
I'm not sure if my remark was clear. I wanted to replace the following:
```
self.cfg: Union[config.Config, None] = None
```
with this which is semantically identical (`Optional[X] := Union[X | None]`) but shorter:
```
self.cfg: Optional[config.Config] = None
```
If you find `Union` more readable, natural I'm OK with it.

sakurai-youhei replied:
@gbanasiak Yeah, the point was clear. I read your comment, I did some research, and I found that the `spam: Optional[X] = None` style seems to be the majority. I will change all to `Optional[config.Config]` accordingly.

sakurai-youhei commented on esrally/driver/driver.py:
```diff
@@ -516,14 +516,14 @@ def receiveMsg_PrepareTrack(self, msg, sender):
         self._seed_tasks(self.processors.get())
         self.send_to_children_and_transition(
-            self, StartTaskLoop(self.track.name, self.cfg), self.Status.INITIALIZING, self.Status.PROCESSOR_RUNNING
+            self, StartTaskLoop(self.track.name, self.cfg), self.Status.INITIALIZING, self.Status.PROCESSOR_RUNNING  # type: ignore[arg-type]  # pylint: disable=C0301  # TODO remove this ignore when introducing type hints
```
Reading the code, it seems to me that `self.cfg` will probably never be None when passed to `StartTaskLoop()`. How about prepending `assert self.cfg is not None` instead of making the argument to `Optional[config.Config]`?

gbanasiak replied:
Agreed, that's better :)

sakurai-youhei replied:
Removed `# type: ignore[arg-type]` by `assert self.cfg is not None` accordingly.

sakurai-youhei commented on esrally/config.py:
```diff
@@ -39,6 +40,138 @@ class Scope(Enum):
+    "tracks",
+    "unit-test",
+]
+Key = Literal[
```
Thanks. Now, sections and keys have reached the expected counts.

sakurai-youhei commented on esrally/racecontrol.py:
```diff
@@ -107,19 +109,26 @@ def receiveUnrecognizedMessage(self, msg, sender):
         self.send(
             self.mechanic,
             mechanic.StartEngine(
-                self.cfg, self.coordinator.metrics_store.open_context, msg.sources, msg.distribution, msg.external, msg.docker
```
Don't know why this line has been left until now, but black 23.3.0 reformats it now.

sakurai-youhei commented:
@gbanasiak Here are the highlights in this round. I appreciate a fresh pair of eyes as this PR gets bigger. Thanks.
- Split out `esrally.types` to avoid circular references between `esrally.config` and other modules.
- Added one more test to check annotation for `cfg` and `config` around functions.
- The number of keys has reached 115, thanks to the test that discovers missing annotations.
- Replaced `Union[T, None]` with `Optional[T]`.
- Eliminated `# pylint: disable=C0301`.

sakurai-youhei commented:
git-rebased in the source branch because this PR gets aged.

gbanasiak reviewed (approved):
Apologies for delays. This looks good to me. Many thanks for this improvement.

The rebase had broken the diff in GH, but https://github.com/elastic/rally/compare/master...sakurai-youhei:literal-check-by-mypy allowed me to see the files that _actually_ changed. I think it might be better to rebase only prior to opening a PR, and then merge master on top of the PR changes.

sakurai-youhei commented:
@gbanasiak I didn't notice my git-rebase messed it up. I reverted it instead of appending something more.

Thank you very much for your review. I've learned a lot through this PR. I will merge this PR once all the CI checks have passed.
